[GUI] Add Sorting Methods to the movie versions/extras folders

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Define sorting methods for content videoassets/videoversions/videoextras. 

The following made sense for versions out of the existing sorting methods:
`label, duration, video resolution, video codec, audio codec, date added, last played, playcount`

for extras. removed video resolution, video codec, audio codec that are not useful.

Created separate "buckets" to store the view type, sort method and order independently for assets versions and extras

### Core Programming
- "last played" and "playcount" override sorting attributes in order to treat folder the same as files, which can result in folders appearing between files rather than always at the top. It could make sense for playlists, tv shows/seasons, but not for the virtual Extras subfolder I think
=> Commit 1 provides a way to bypass the override.

- "duration", "video resolution", "video codec" and "audio codec" sorting methods require the stream details to be loaded to work properly for video assets, but they're not for movies right after loading the data with GetMovieByWhere. They are loaded shortly afterwards though (as part of the background artwork loader?)
=> Commit 3 fixes by forcing the stream details retrieval as part of the item load from database (only for assets).
An alternative would a wait for the async background load to complete, but how is performance (any parallelism?)
> Opinions?

Note: both make individual stream details queries  - not optimal but OK for now with the relatively short lists of versions or extras of a single movie.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Now that versions and extras are presented as folders, let's sort them in interesting ways!

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

went through the different sorting methods for versions and extras.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

Ability to sort versions and extras in different ways than the asset creation order.

## Screenshots (if appropriate):
Versions:
![image](https://github.com/user-attachments/assets/da435fc2-74aa-4b4f-bc97-22ff4695a35b)

Extras:
![image](https://github.com/user-attachments/assets/48147aa7-d13d-4344-9b40-a75ac506622f)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP committed:
[video] Always retrieve stream details of movie assets

Automatic stream details loading is too late to be reliably used in sorting before display with GUIWindowVideoBase and children.

Without stream details, duration is retrieved from movie metadata and does not reflect the variable durations of the different versions/extras of the movie.

WIP: alternative would be to wait for the art/tag info loading before proceeding.

CrystalP committed:
[SortUtils] Add SortAttribute to force folders to be treated as such

CrystalP committed:
[GUI] Add sort methods for assets/versions/extras

jjd-uk commented:
For me populating label2 from the skin side is very easy, you simply need:

```
	<variable name="ListLabel2Var">
		<value condition="String.IsEmpty(Container.PluginName) + Container.Content(tvshows) + Container.SortMethod(29)">$VAR[WatchedStatusVar]</value>
		<value condition="String.IsEmpty(Container.PluginName) + String.IsEqual(ListItem.DBType,season) + Container.SortMethod(29)">$VAR[WatchedStatusVar]</value>
		<value condition="String.IsEmpty(Container.PluginName) + String.IsEqual(ListItem.DBType,set) + Container.SortMethod(1)">$VAR[WatchedStatusVar]</value>
		<value condition="ListItem.IsFolder + [Window.IsActive(TVRecordings) | Window.IsActive(RadioRecordings)]">$VAR[WatchedStatusVar]</value>
		<value condition="Container.SortMethod(7) | Container.SortMethod(29)">$INFO[ListItem.Year]</value>
		<value condition="!String.isempty(ListItem.Appearances)">$LOCALIZE[38026]: $INFO[ListItem.Appearances]</value>
		<value condition="Window.IsActive(musicplaylist) | Window.IsActive(videoplaylist)">$INFO[ListItem.Duration]</value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(1) | Container.SortMethod(9)"></value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(32)">$INFO[ListItem.VideoResolution]$VAR[VideoResolutionTypeVar, ]</value>#
		<value condition="$EXP[videoasset_true] + Container.SortMethod(33)">$VAR[VideoCodecVar]</value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(36)">$VAR[AudioCodecVar]$VAR[ObjectAudioTypeVar, / ]</value>
		<value>$INFO[ListItem.Label2]</value>
	</variable>
```

at https://github.com/xbmc/xbmc/blob/master/addons/skin.estuary/xml/Variables.xml#L91-L100

As you can see we already do this since label2 does not always provide what we want in the skin.

ksooo commented on xbmc/filesystem/VideoDatabaseDirectory/DirectoryNodeMovieAssets.cpp:
```diff
@@ -41,7 +41,7 @@ bool CDirectoryNodeMovieAssets::GetContent(CFileItemList& items) const
 
   const int details{items.HasProperty("set_videodb_details")
                         ? items.GetProperty("set_videodb_details").asInteger32()
-                        : VideoDbDetailsNone};
+                        : VideoDbDetailsStream};
```
This could be simplified to:

```c++
  const int details{items.GetProperty("set_videodb_details").asInteger32(VideoDbDetailsStream)};
```

CrystalP committed:
[Estuary] Provide label2 values not available from core for video asset content - thanks jjd-uk

- Also fix missing space for "View type" in left bar
- And fix missing duration label2 for List, Widelist, Fanart views

CrystalP commented:
> As you can see we already do this since label2 does not always provide what we want in the skin

OK, that can work for now until the dust settles and a sort by quality appears (or not!).
Snippet incorporated in a commit

I went ahead and removed the sorting methods that don't make a lot of sense for extras (video res, video and audio codec), but left them for versions.

Also addressed the code comment, thanks ksooo.

ksooo reviewed (approved):
core code changes look good to me

jjd-uk commented:
To resolved duration not appearing when `Sort by: Duration` is selected in views other than `MediaList`  adjust `ListLabel2Var` to 

```
	<variable name="ListLabel2Var">
		<value condition="String.IsEmpty(Container.PluginName) + Container.Content(tvshows) + Container.SortMethod(29)">$VAR[WatchedStatusVar]</value>
		<value condition="String.IsEmpty(Container.PluginName) + String.IsEqual(ListItem.DBType,season) + Container.SortMethod(29)">$VAR[WatchedStatusVar]</value>
		<value condition="String.IsEmpty(Container.PluginName) + String.IsEqual(ListItem.DBType,set) + Container.SortMethod(1)">$VAR[WatchedStatusVar]</value>
		<value condition="ListItem.IsFolder + [Window.IsActive(TVRecordings) | Window.IsActive(RadioRecordings)]">$VAR[WatchedStatusVar]</value>
		<value condition="Container.SortMethod(7) | Container.SortMethod(29)">$INFO[ListItem.Year]</value>
		<value condition="!String.isempty(ListItem.Appearances)">$LOCALIZE[38026]: $INFO[ListItem.Appearances]</value>
		<value condition="Window.IsActive(musicplaylist) | Window.IsActive(videoplaylist)">$INFO[ListItem.Duration]</value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(1)"></value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(9) + Control.IsVisible(504)"></value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(9) + !Control.IsVisible(504)">$INFO[ListItem.Duration]</value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(32)">$INFO[ListItem.VideoResolution]$VAR[VideoResolutionTypeVar, ]</value>#
		<value condition="$EXP[videoasset_true] + Container.SortMethod(33)">$VAR[VideoCodecVar]</value>
		<value condition="$EXP[videoasset_true] + Container.SortMethod(36)">$VAR[AudioCodecVar]$VAR[ObjectAudioTypeVar, / ]</value>
		<value condition="Container.SortMethod(9)">$INFO[ListItem.Duration]</value>
		<value>$INFO[ListItem.Label2]</value>
	</variable>
```

CrystalP commented:
Thanks. Is the # at the end of the 
`<value condition="$EXP[videoasset_true] + Container.SortMethod(32)">$INFO[ListItem.VideoResolution]$VAR[VideoResolutionTypeVar, ]</value>#` line on purpose?

jjd-uk commented:
No the # must a c&p error

jjd-uk reviewed (approved):
LGTM from a functional pov

CrystalP commented:
force-pushed the typo fix
Also dug deeper in core to understand the mechanism to store/restore/default the view type/sort method and order while navigating, and after a change to separate storage for assets, versions and extras I think this works now rather well, addressing my original concerns. That required a minor change to commit 1.

Hitcher commented:
One thing I'm not keen on is the Extras item isn't listed at the end of the Versions list now.

CrystalP commented:
> One thing I'm not keen on is the Extras item isn't listed at the end of the Versions list now.

Me neither, but that's the standard way folders get sorted in Kodi. It is possible to add a new flag for a sort that moves folders to the bottom.
Since the default version is selected by default when opening the movie folder, it's not really a usability issue though I think.
